mysys/my_largepage.c

```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
+  }
+  else
+  {
+    while (i < my_large_page_sizes_length &&
+          (r= readdir(dirp)))
+    {
+      if (strncmp("hugepages-", r->d_name, 10) == 0)
+      {
+        sizes[i]= strtoull(r->d_name + 10, NULL, 10) * 1024ULL;
+        if (!my_is_2pow(sizes[i]))
+        {
+          fprintf(stderr, "Warning: non-power of 2 large page size (%zu) found, skipping\n", sizes[i]);
+          sizes[i] = 0;
+          continue;
+        }
+        ++i;
+      }
+    }
+    if (closedir(dirp))
+    {
+      perror("Warning: failed to close /sys/kernel/mm/hugepages");
+    }
+    qsort(sizes, i, sizeof(size_t), size_t_cmp);
+  }
+  DBUG_VOID_RETURN;
+}
+#endif
+
+/* Multisized (Linux/FreeBSD) large pages allocator  */
+
+#if defined(HAVE_LINUX_LARGE_PAGES) || defined(HAVE_MMAP_ALIGNED)
+uchar* my_large_malloc_int(size_t *size, myf my_flags)
 {
-  int shmid;
   uchar* ptr;
-  struct shmid_ds buf;
+  int mapflag;
+  int page_i= 0;
+  size_t large_page_size= 0;
   DBUG_ENTER("my_large_malloc_int");
 
-  /* Align block size to my_large_page_size */
-  size= MY_ALIGN(size, (size_t) my_large_page_size);
-  
-  shmid = shmget(IPC_PRIVATE, size, SHM_HUGETLB | SHM_R | SHM_W);
-  if (shmid < 0)
+  while (1)
   {
-    if (my_flags & MY_WME)
-      fprintf(stderr,
-              "Warning: Failed to allocate %lu bytes from HugeTLB memory."
-              " errno %d\n", (ulong) size, errno);
+    mapflag= MAP_PRIVATE | MAP_ANONYMOUS;
+    if (my_use_large_pages)
+    {
+      large_page_size= my_next_large_page_size(*size, &page_i);
+      if (large_page_size)
+      {
+#ifdef HAVE_LINUX_LARGE_PAGES
+        mapflag|= MAP_HUGETLB | my_bit_log2(large_page_size) << MAP_HUGE_SHIFT;
+#else
+        mapflag|= MAP_ALIGNED_SUPER | MAP_ALIGNED(my_bit_log2(large_page_size));
+#endif
+      }
+    }
+    /* mmap adjusts the size to the hugetlb page size so no adjustment is needed */
+    ptr= mmap(NULL, *size, PROT_READ | PROT_WRITE, mapflag, -1, 0);
+    if (ptr == (void*) -1)
+    {
+      ptr= NULL;
+      if (my_flags & MY_WME)
+      {
+        if (large_page_size)
+        {
+          fprintf(stderr,
+                  "Warning: Failed to allocate %zu bytes from HugeTLB memory"
+                  "(page size %zu). errno %d\n", *size, large_page_size, errno);
+        }
+        else
+        {
+          fprintf(stderr,
+                  "Warning: Failed to allocate %zu bytes from memory."
+                  " errno %d\n", *size, errno);
+        }
+      }
+      /* try next smaller memory size */
+      if (large_page_size && errno == ENOMEM)
+        continue;
 
-    DBUG_RETURN(NULL);
+      /* other errors are more serious */
+      DBUG_RETURN(NULL);
+    }
+    else /* success */
+    {
+      if (large_page_size)
+      {
+         /* we do need to record the adjustment so that munmap gets called with
+            the right size. This is only the case for HUGETLB pages.
+            Align block size to large_page_size */
```

Comment:
```
/*
  we do need to record the adjustment so that munmap gets called with
  the right size. This is only the case for HUGETLB pages.
  Align block size to large_page_size
*/
```